src/Microsoft.AspNet.Diagnostics/ErrorPageMiddleware.cs

```diff
@@ -182,31 +188,103 @@ private StackFrame StackFrame(Chunk line)
+                    fileProvider: _options.FileProvider,
+                    sourceCodeLineCount: _options.SourceCodeLineCount);
+            }
+            else
+            {
+                return GetStackFrame(function, file, lineNumber, _options.FileProvider, _options.SourceCodeLineCount);
+            }
         }
 
-        private StackFrame LoadFrame(string function, string file, int lineNumber)
+        internal static StackFrame GetStackFrame(
+            string function,
+            string file,
+            int lineNumber,
+            IFileProvider fileProvider,
+            int sourceCodeLineCount)
         {
             var frame = new StackFrame { Function = function, File = file, Line = lineNumber };
+
+            if (string.IsNullOrEmpty(file))
+            {
+                return frame;
+            }
+
+            IEnumerable<string> lines = null;
             if (File.Exists(file))
             {
-                var code = File.ReadLines(file);
-                ReadFrameContent(frame, code, lineNumber, lineNumber);
+                lines = File.ReadLines(file);
             }
+            else
+            {
+                // Handle relative paths and embedded files
+                var fileInfo = fileProvider.GetFileInfo(file);
+                if (fileInfo.Exists)
+                {
+                    // ReadLines doesn't accept a stream. Use ReadLines as its more efficient
+                    // relative to reading lines via stream reader
+                    if (!string.IsNullOrEmpty(fileInfo.PhysicalPath))
+                    {
+                        lines = File.ReadLines(fileInfo.PhysicalPath);
+                    }
+                    else
+                    {
+                        lines = GetLines(fileInfo.CreateReadStream());
+                    }
+                }
+            }
+
+            if (lines != null)
+            {
+                GetErrorCodeBlock(frame, lines, lineNumber, sourceCodeLineCount);
+            }
+
             return frame;
         }
 
-        private void ReadFrameContent(StackFrame frame,
-                                      IEnumerable<string> code,
-                                      int startLineNumber,
-                                      int endLineNumber)
+        private static void GetErrorCodeBlock(
+            StackFrame frame,
+            IEnumerable<string> allLines,
+            int errorLineNumberInFile,
+            int sourceCodeLineCount)
+        {
+            // Get the line boundaries in the file to be read and read all these lines at once into an array.
+            // This will cause the backing stream to be closed.
+            var preErrorLineNumberInFile = Math.Max(errorLineNumberInFile - sourceCodeLineCount, 1);
+            var postErrorLineNumberInFile = errorLineNumberInFile + sourceCodeLineCount;
+            var codeBlock = allLines
+                .Skip(preErrorLineNumberInFile - 1)
```

Comment:
is the `-1` purely for the case that `errorLineNumberInFile` > `sourceCodeLineCount`? Or is it a magic number? If it's the first, just make the `Math.Max` do `0` instead of `1`